SpatialGDK/ExternalSchemaCodegen/Programs/Improbable.CodeGen.Unreal/HeaderGenerator.cs

```diff
@@ -0,0 +1,376 @@
+            builder.AppendLine($@"~Update() = default;");
+            builder.AppendLine($@"bool operator==(const Update&) const;");
+            builder.AppendLine($@"bool operator!=(const Update&) const;");
+            builder.Append(Environment.NewLine);
+            builder.AppendLine($@"// Creates an Update from a {qualifiedType} object.");
+            builder.AppendLine($@"static Update FromInitialData(const {qualifiedType}& Data);");
+            builder.Append(Environment.NewLine);
+            builder.AppendLine($@"/**");
+            builder.AppendLine($@"  * Converts to a {qualifiedType}");
+            builder.AppendLine($@"  * object. It is an error to call this function unless *all* of the optional fields in this");
+            builder.AppendLine($@"  * update are filled in.");
+            builder.AppendLine($@"  */");
+            builder.AppendLine($@"{qualifiedType} ToInitialData() const;");
+            builder.Append(Environment.NewLine);
+            builder.AppendLine($@"/**");
+            builder.AppendLine($@"  * Replaces fields in the given {qualifiedType}");
+            builder.AppendLine($@"  * object with the corresponding fields in this update, where present.");
+            builder.AppendLine($@"  */");
+            builder.AppendLine($@"void ApplyTo({qualifiedType}&) const;");
+            builder.Append(Environment.NewLine);
+            builder.AppendLine($@"// Serialize this update object data into the C API component update argument");
+            builder.AppendLine($@"void Serialize(Schema_ComponentUpdate* ComponentUpdate) const override;");
+            builder.Append(Environment.NewLine);
+            builder.AppendLine($@"// Deserialize the C API component update argument into an instance of this class and return it");
+            builder.AppendLine($@"static Update Deserialize(Schema_ComponentUpdate* ComponentUpdate);");
+
+            return builder.ToString();
+        }
+
+        private static string GenerateOptionalFieldAccessorsDeclarations(string name, TypeDescription type, Bundle bundle)
+        {
+            var fieldText = new StringBuilder();
+
+            foreach (var field in type.Fields)
+            {
+                var qualifiedFieldType = Types.GetFieldTypeAsCpp(field, bundle);
+                fieldText.AppendLine($"// Field {Text.SnakeCaseToPascalCase(field.Identifier.Name)} = {field.FieldId}");
+                fieldText.AppendLine($"const {Types.CollectionTypesToQualifiedTypes[Types.Collection.Option]}<{qualifiedFieldType}>& Get{Text.SnakeCaseToPascalCase(field.Identifier.Name)}() const;");
+                fieldText.AppendLine($"{Types.CollectionTypesToQualifiedTypes[Types.Collection.Option]}<{qualifiedFieldType}>& Get{Text.SnakeCaseToPascalCase(field.Identifier.Name)}();");
+                fieldText.AppendLine($"{name}::Update& Set{Text.SnakeCaseToPascalCase(field.Identifier.Name)}({qualifiedFieldType});");
```

Comment:
Why are we returning the update object when calling the set function, doesn't it mutate the state of the update object it is invoked on?

Reply:
it does, but the constructor pattern kinda works that way, i.e. 

`auto MyUpdate = new Position.Update().SetCoords(1,2,3)`